// Generated by CoffeeScript 1.4.0
var AssertionError, Realizer;

AssertionError = require('assert').AssertionError;

require('fing');

Realizer = (function() {

  function Realizer() {}

  Realizer.realizers = {};

  Realizer.createFunction = function(name, object, configuration, realization) {
    var key;
    key = object.fing.ref + ':' + name;
    if (typeof this.realizers[key] === 'undefined') {
      this.realizers[key] = {
        object: object,
        realizations: []
      };
      object[name] = function() {
        var realizerCallback;
        object = Realizer.realizers[key].object;
        realizerCallback = Realizer.realizers[key].realizations.shift();
        if (typeof realizerCallback === 'undefined') {
          throw new AssertionError({
            message: 'Unexpected call to ' + key
          });
        }
        return realizerCallback({
          object: object,
          args: arguments
        });
      };
    }
    return this.realizers[key].realizations.push(realization);
  };

  return Realizer;

})();

module.exports = Realizer;
